### DISCUSSION TITLE
HTML Forms in Nuxt Components

### DISCUSSION THREAD

**Discussion by mubaidr** [nuxt/nuxt#23548] [Ideas]
I hope you're all doing well. I've been using Nuxt 3 for my web development projects, and I've come across a feature in Astro that I find extremely valuable - the ability to handle forms in SSR mode without sending JavaScript to the client. This feature simplifies the development process and improves performance significantly. More details here: https://docs.astro.build/en/recipes/build-forms/

Considering that we have experimental support for Server-side components (https://nuxt.com/docs/api/components/nuxt-island), this could go hand in hand. 

I wanted to discuss the possibility of having a similar feature in Nuxt 3. Currently, Nuxt 3 is an excellent framework for SSR, but handling forms on the server side without relying on client-side JavaScript can be a bit challenging. Right now, the only solution is building API.

Here are a few reasons why I believe this feature would be beneficial: Performance, Simplicity (even the validation part), Security, Compatibility: Some users may have JavaScript disabled, so having a server-side fallback for form submissions ensures broad compatibility.

I understand that Nuxt 3 is already a powerful framework, but adding this feature could make it even more versatile and developer-friendly. It would be great if we could discuss the feasibility of implementing SSR form handling in Nuxt 3 and gather insights from the community on how this could be achieved.

What are your thoughts on this? Do you think it's a valuable feature to have in Nuxt 3, and if so, how can we work together to make it happen?

**Comment by huang-julien** [comment]
Hi :wave: Maybe can you try https://github.com/Hebilicious/form-actions-nuxt from the great @Hebilicious :D

**Reply by mubaidr** [reply]
This is good effort would be very useful. But I am thinking about calling typed functions from templates.

Simplest explaination: 
- define a typesafe function in script section
- call this function by name in template. Either in script or as form action.

Basically like astro Weds trpc like behavior.

**Reply by Hebilicious** [reply]
FYI If you want to define server codes or form actions within your SFC, you can use this module https://github.com/Hebilicious/server-block-nuxt

**Reply by mubaidr** [reply]
I recommend you check astro documentatiom (I have linked above) to get an idea what this discussion is about and how it feels much better to use IMHO. 

P.S. Server blocks are good to combine backend code into one place.

**Reply by Hebilicious** [reply]
@mubaidr I recommend you try out these modules to get an idea of how they work. They allow you to do something which is very similar to Astro. Nuxt is open source, you are very welcome to contribute suggestions and feedback on these existing modules, or to create your own / opening a PR.

```html
<server>
export const POST = defineFormActions({
  signIn: async (event) => {
    const formData = await readFormData(event)
    // ...
  }
})
</server>
<script server>
const { enhance, data } = await useFormAction()
</script>
<template>
  <form v-enhance="enhance" method="POST" action="login">
    <pre> {{ data }} </pre>
    <label>
      Email
      <input name="email" type="email" :value="data.formResponse?.email ?? ''">
    </label>
    <label>
      Password
      <input name="password" type="password">
    </label>
    <button>Log in</button>
  </form>
</template>
```

**Reply by blocka** [reply]
@Hebilicious I'm unable to get this working with server blocks...as far as I can tell it doesn't generate the correct code in .generated/.loader

Do you by chance have an example up somewhere on stackblitz or codesandbox?

**Reply by Hebilicious** [reply]
You can find some examples there :

https://github.com/Hebilicious/form-actions-nuxt-examples/blob/advanced/pages/mangas/%5Bid%5D.vue

**Reply by blocka** [reply]
![image](https://github.com/nuxt/nuxt/assets/43194/b81e5668-947b-4f09-b465-c39a85ab518d)
this also seems to have issues?

**Reply by Hebilicious** [reply]
Could be related to your deps/local setup...
I would recommend you fork either one of the repo and run the playgrounds.
Make sure you are using the same Nuxt/vite version as in the lockfile.

**Reply by blocka** [reply]
I forked _this_ repo. The only difference is that I used pnpm not bun.
Should that make a difference?

**Reply by Hebilicious** [reply]
It will yes, the lockfiles are not the same. Actually there's no lockfiles in the example-repo, so you can try the playgrounds from the main repo just to be sure... The example repo auto-extract the playgrounds.
Sorry about the hassle, I'll investigate this.
Looks like the server-block-nuxt works fine though, it generates the loader correctly. It appears to be something nuxt related that writes the wrong path, it should work fine with earlier versions of nuxt.
